### PR TITLE
bpo-1635741: Port sha256 module to multiphase init (PEP 489)

### PR DESCRIPTION
**Pull request by koubaa**
[Link to issue](https://bugs.python.org/issue1635741)

https://bugs.python.org/issue1635741

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@koubaa

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by koubaa** [comment]
edtited the commit msg and repushed so that it would pick up the CLA which I signed yesterday.  Not sure what the Misc/NEWS.d/next is about

File: Modules/sha256module.c [hunk 1]
```diff
@@ -681,43 +681,44 @@ static struct PyMethodDef SHA_functions[] = {
...
+        return -1;
     }
 
-    m = PyModule_Create(&_sha256module);
-    if (m == NULL)
-        return NULL;
-
     Py_INCREF((PyObject *)&SHA224type);
-    PyModule_AddObject(m, "SHA224Type", (PyObject *)&SHA224type);
+    PyModule_AddObject(module, "SHA224Type", (PyObject *)&SHA224type);
```
Comment (shihai1991) [inline Modules/sha256module.c]: `Unlike other functions that steal references, PyModule_AddObject() only decrements the reference count of value on success.` from https://docs.python.org/3/c-api/module.html?highlight=pymodule_addobject#c.PyModule_AddObject

so you need do something like:
```
Py_INCREF((PyObject *)&SHA224type);
if (PyModule_AddObject(module, "SHA224Type", (PyObject *)&SHA224type) < 0) {
    Py_DECREF(PyObject *)&SHA224type);
    return -1;
}
```

File: Modules/sha256module.c [hunk 2]
```diff
@@ -681,43 +681,44 @@ static struct PyMethodDef SHA_functions[] = {
...
+    PyModule_AddObject(module, "SHA256Type", (PyObject *)&SHA256type);
+    return 0;
+}
+
+static PyModuleDef_Slot _sha256_slots[] = {
+    {Py_mod_exec, sha256_exec},
+    {0, NULL}
+};
 
+static struct PyModuleDef _sha256module = {
```
Comment (shihai1991) [inline Modules/sha256module.c:713]: designated initializers (especially nice for type declarations) from pep7: https://www.python.org/dev/peps/pep-0007/

something like:
https://github.com/python/cpython/blob/master/Modules/_gdbmmodule.c#L761-L771

File: Modules/sha256module.c [hunk 3]
```diff
@@ -681,43 +681,44 @@ static struct PyMethodDef SHA_functions[] = {
...
-    if (m == NULL)
-        return NULL;
-
     Py_INCREF((PyObject *)&SHA224type);
-    PyModule_AddObject(m, "SHA224Type", (PyObject *)&SHA224type);
+    PyModule_AddObject(module, "SHA224Type", (PyObject *)&SHA224type);
     Py_INCREF((PyObject *)&SHA256type);
-    PyModule_AddObject(m, "SHA256Type", (PyObject *)&SHA256type);
-    return m;
+    PyModule_AddObject(module, "SHA256Type", (PyObject *)&SHA256type);
```
Comment (shihai1991) [inline Modules/sha256module.c]: MAYBE you can move those typeobject to module state(you can use a separate PR to do this job).

something like this PR: https://github.com/python/cpython/commit/ec689187957cc80af56b9a63251bbc295bafd781

**Comment by shihai1991** [comment]
> edtited the commit msg and repushed so that it would pick up the CLA which I signed yesterday. Not sure what the Misc/NEWS.d/next is about

If you add a feature which could effect the user, you should describe what you have changed in Misc/NEWS.d/next.
you can use a web service or cli to generate the description info:
* https://blurb-it.herokuapp.com/
* https://pypi.org/project/blurb/

A reference PR of Dong-hee Na:
https://github.com/python/cpython/commit/c4862e333ab405dd5789b4061222db1982147de4

File: Modules/sha256module.c (hunk 1 above)
Comment (koubaa) [reply Modules/sha256module.c]: @shihai1991 does this mean there was a possible leak pre-existing before my changes?

Comment (shihai1991) [reply Modules/sha256module.c]: Yes, you are right. you can fix this leak in your PR ;)

File: Modules/sha256module.c (hunk 3 above)
Comment (vstinner) [reply Modules/sha256module.c]: It's ok to convert static types to heap types in a separated PR.

**Review by vstinner — COMMENTED** [review]
cc @corona10

File: Modules/sha256module.c [hunk 4]
```diff
@@ -681,43 +681,44 @@ static struct PyMethodDef SHA_functions[] = {
...
+}
+
+static PyModuleDef_Slot _sha256_slots[] = {
+    {Py_mod_exec, sha256_exec},
+    {0, NULL}
+};
 
+static struct PyModuleDef _sha256module = {
+    PyModuleDef_HEAD_INIT,
+    "_sha256",
```
Comment (corona10) [inline Modules/sha256module.c]: I'd like to change this as the C99 style.
```suggestion
    .m_name = "_sha256",
    .m_size = 0,
    .m_methods = SHA_functions,
    .m_slots = _sha256_slots
```

**Review by corona10 — CHANGES_REQUESTED** [review]

**Review by corona10 — COMMENTED** [review]
> It's ok to convert static types to heap types in a separated PR.

If @vstinner is okay to make a separate PR for heap types.
I am also okay to merge this PR.

Please apply the review :)

File: Modules/sha256module.c (hunk 2 above)
Comment (koubaa) [reply Modules/sha256module.c:713]: @shihai1991 you're right, this is more clear.

File: Modules/sha256module.c (hunk 3 above)
Comment (koubaa) [reply Modules/sha256module.c]: @shihai1991 What is the advantage of that?  Does this help with subintepreter issues?

Comment (koubaa) [reply Modules/sha256module.c]: @shihai1991 oh, I think I get it.  static ptrs are global state that can not be isolated to a given subinterpreter.  I can give it a shot but maybe in another PR if I can't figure out how right away

File: Modules/sha256module.c [hunk 5]
```diff
@@ -681,43 +681,50 @@ static struct PyMethodDef SHA_functions[] = {
...
+static struct PyModuleDef _sha256module = {
+    PyModuleDef_HEAD_INIT,
+    .m_name = "_sha256",
+    .m_doc = NULL,
+    .m_size = 0,
+    .m_methods = SHA_functions,
+    .m_slots = _sha256_slots,
+    .m_traverse = NULL,
+    .m_clear = NULL,
+    .m_free = NULL
```
Comment (vstinner) [inline Modules/sha256module.c]: Please remove members which are set to NULL: it's already done by default. You can also remove .m_size=0.

**Comment by koubaa** [comment]
@corona10 any change still requested?

@shihai1991 @vstinner I started looking at doing heap types.  First a disclaimer I'm new in this code so maybe I'm missing some context.  I am seeing auto-generated files from clinic that don't take a module pointer as an input, so the module->state approach used by _gdbmmodule would only work if I could modify all of those auto-generated files (and presumably the system which generates them).

**Review by vstinner — APPROVED** [review]
LGTM.

@corona10, @shihai1991: Would you mind to review the change?

@corona10: Feel free to merge if the change looks good to you ;-)

**Review by corona10 — APPROVED** [review]
LGTM Thanks for the work!

**Comment by shihai1991** [comment]
oh, sorry, I am late, LGTM.

**Pull request merged**